### ISSUE TITLE
Unclear output for Pytorch Profiler

### ISSUE THREAD

**Issue by ralzq01** [pytorch/pytorch#33855]
## 📚 Documentation on PyTorch Profiler

I don't fully understand the output of [PyTorch Profiler](https://pytorch.org/docs/stable/autograd.html#torch.autograd.profiler.profile) after using `torch.autograd.profiler` (combined with `torch.autograd.profiler.record_function` for labeling). To be specifically, the output of `prof.key_averages().table(sort_by="cuda_time_total")` gets the following table:

| Name | Self CPU total % | Self CPU total | CPU total % | CPU total | CPU time avg | CUDA total % | CUDA total | CUDA time avg | Number of Calls |
| ---- | ---------------- | -------------- | ----------- | --------- | ------------ | ------------ | ---------- | ------------- | --------------- |
|      |                  |                |             |           |              |              |            |               |                 |

It seems the document doesn't explain much for the clear meanings of each item. To be concretely, I will very appreciate if the document can provide answers to the following questions:

1. What's the meaning and the difference of `Self CPU total` and `CPU total`
2. Is the `CUDA total` purely the execution time for the cuda code? Or it also includes time of kernel launching, kernel queueing in GPU? 
3. How can I understand the `CPU total` for operations which are executed on GPU, e.g., model forwarding and backwarding. Since these operations are done in GPU, but I still find even unneglectable time cost spent by CPU.

**Comment by ailzhang** [comment]
It'll definitely help if we could improve the documentation there.

**Comment by bytp6** [comment]
If Conv2d is used in the model, in the profiler, it shows CUDA_total_time of conv2d, convolution, _convolution, and cudnn_convolution. I wonder that what's the meaning and the difference between them.

**Comment by ezyang** [comment]
They're internal wrappers that call each other and do some processing before deferring to the next implementation. You don't have to worry about it; really cudnn_convolution (or the top level conv2d) is all you really care about.

**Comment by ralzq01** [comment]
@ezyang Thanks for your reply! Could you also offer some hints to the above questions? Particularly, I want to know the accurate meaning of CUDA time. Does it also include the kernel launching time and queueing time?

**Comment by ezyang** [comment]
Sorry, I don't know off the top of my head. If I needed to figure it out I'd go look and see where the times are coming from in the source code. Be sure to report back here when you find out :)

**Comment by WangDeyao** [comment]
Hi!

Following the previous discussion, I also spot a potential problem.

When calculating the "CUDA time total" in the end, it seems that conv2d, convolution, _convolution has all been added up. This makes the "CUDA time total" too big. A conv2d will essentially been added 3 times (4 times?). Other operations might also been counted multiple times according to how the functions are wrapped. What's more, this also leaves CUDA time % inaccurate.

Here is an example of 
```
print(prof.key_averages().table(sort_by='cuda_time_total'))
```

--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                        Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
conv2d                      1.80%            3.172ms          21.41%           37.707ms         200.571us        20.75%           127.768ms        679.615us        188              
convolution                 2.76%            4.854ms          19.60%           34.536ms         183.700us        20.62%           126.947ms        675.251us        188              
_convolution                3.52%            6.206ms          16.85%           29.681ms         157.880us        20.29%           124.933ms        664.537us        188              
cudnn_convolution           7.53%            13.263ms         7.53%            13.263ms         70.550us         17.71%           109.017ms        579.880us        188              
cat                         5.14%            9.059ms          5.14%            9.059ms          62.478us         2.62%            16.101ms         111.044us        145              
add_                        2.87%            5.051ms          2.87%            5.051ms          26.867us         1.98%            12.182ms         64.796us         188              
batch_norm                  0.49%            868.843us        3.19%            5.625ms          187.506us        1.39%            8.577ms          285.901us        30               
_batch_norm_impl_index      1.15%            2.029ms          2.70%            4.756ms          158.544us        1.37%            8.461ms          282.043us        30               
mul                         2.63%            4.636ms          2.63%            4.636ms          25.470us         1.18%            7.269ms          39.940us         182              
relu_                       0.76%            1.337ms          0.76%            1.337ms          26.749us         1.15%            7.067ms          141.334us        50               
div                         2.41%            4.239ms          2.41%            4.239ms          28.449us         1.14%            7.007ms          47.029us         149              
add                         1.63%            2.876ms          1.63%            2.876ms          29.960us         1.10%            6.794ms          70.771us         96               
instance_norm               0.27%            483.988us        1.29%            2.267ms          151.121us        1.07%            6.590ms          439.364us        15               
native_batch_norm           0.43%            756.114us        0.43%            756.114us        50.408us         1.00%            6.167ms          411.102us        15               
matmul                      0.04%            66.167us         0.13%            232.600us        232.600us        0.84%            5.189ms          5.189ms          1                
bmm                         0.03%            48.289us         0.03%            48.289us         48.289us         0.82%            5.042ms          5.042ms          1                
sub                         1.89%            3.334ms          1.89%            3.334ms          27.331us         0.62%            3.824ms          31.348us         122              
grid_sampler                0.33%            584.806us        1.18%            2.087ms          43.477us         0.51%            3.143ms          65.472us         48               
cudnn_grid_sampler          0.85%            1.502ms          0.85%            1.502ms          31.294us         0.48%            2.945ms          61.355us         48               
reshape                     1.37%            2.407ms          2.70%            4.758ms          20.162us         0.35%            2.142ms          9.077us          236              
relu                        0.90%            1.590ms          0.90%            1.590ms          26.071us         0.33%            2.047ms          33.557us         61               
to                          50.89%           89.649ms         51.24%           90.260ms         1.612ms          0.33%            2.043ms          36.480us         56               
avg_pool2d                  0.06%            110.242us        0.06%            110.242us        27.560us         0.31%            1.916ms          478.975us        4                
view                        2.53%            4.450ms          2.53%            4.450ms          9.509us          0.31%            1.905ms          4.071us          468              
cudnn_batch_norm            0.56%            987.281us        0.56%            987.281us        65.819us         0.29%            1.816ms          121.042us        15               
meshgrid                    1.05%            1.846ms          2.03%            3.575ms          71.492us         0.24%            1.486ms          29.729us         50               
sigmoid                     0.73%            1.286ms          0.73%            1.286ms          26.792us         0.23%            1.443ms          30.060us         48               
stack                       0.93%            1.633ms          0.93%            1.633ms          32.664us         0.20%            1.258ms          25.155us         50               
tanh                        0.42%            732.950us        0.42%            732.950us        29.318us         0.12%            759.754us        30.390us         25               
rsub                        0.48%            847.420us        0.48%            847.420us        35.309us         0.11%            705.594us        29.400us         24               
contiguous                  1.36%            2.401ms          1.37%            2.422ms          8.650us          0.11%            698.429us        2.494us          280              
upsample_bilinear2d         0.31%            552.146us        0.31%            552.146us        46.012us         0.11%            665.609us        55.467us         12               
empty_strided               0.35%            611.458us        0.35%            611.458us        11.537us         0.08%            503.859us        9.507us          53               
expand                      0.55%            971.620us        0.55%            971.620us        9.526us          0.06%            389.244us        3.816us          102              
split_with_sizes            0.39%            691.426us        0.39%            691.426us        13.829us         0.06%            345.447us        6.909us          50               
Scatter                     0.07%            123.146us        0.11%            189.578us        94.789us         0.03%            188.416us        94.208us         2                
repeat                      0.07%            122.235us        0.07%            122.235us        61.118us         0.01%            61.766us         30.883us         2                
permute                     0.16%            284.320us        0.16%            284.320us        11.847us         0.01%            59.035us         2.460us          24               
chunk                       0.02%            31.036us         0.03%            55.954us         27.977us         0.01%            54.272us         27.136us         2                
detach                      0.05%            82.179us         0.05%            82.179us         6.848us          0.01%            45.590us         3.799us          12               
empty                       0.07%            127.385us        0.07%            127.385us        7.962us          0.01%            31.736us         1.983us          16               
split                       0.01%            24.918us         0.01%            24.918us         12.459us         0.00%            25.600us         12.800us         2                
unsqueeze                   0.01%            24.563us         0.01%            24.563us         12.281us         0.00%            14.500us         7.250us          2                
bernoulli_                  0.08%            147.275us        0.08%            147.275us        73.638us         0.00%            11.266us         5.633us          2                
empty_like                  0.01%            10.421us         0.01%            21.051us         21.051us         0.00%            5.119us          5.119us          1                
_unsafe_view                0.01%            9.058us          0.01%            9.058us          9.058us          0.00%            3.070us          3.070us          1                
sqrt                        0.01%            22.307us         0.01%            22.307us         22.307us         0.00%            2.047us          2.047us          1                
transpose                   0.01%            10.394us         0.01%            10.394us         10.394us         0.00%            2.045us          2.045us          1                
detach_                     0.00%            5.105us          0.00%            5.105us          5.105us          0.00%            1.027us          1.027us          1                
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 176.159ms
CUDA time total: 615.652ms
